src/inmanta/protocol/endpoints.py

```diff
@@ -474,3 +477,65 @@ async def _call(
 
         result = await self._transport_instance.call(method_properties, args, kwargs)
         return result
+
+
+class TypedClient(Client):
+    """A client that returns typed data instead of JSON"""
+
+    def _raise_exception(self, exception_class: type[exceptions.BaseHttpException], result: types.JsonType) -> None:
+        """Raise an exception based on the provided status"""
+        message = result.get("message", None)
+        details = result.get("error_details", None)
+
+        raise exception_class(message, details)
+
+    def _process_response(self, method_properties: common.MethodProperties, response: common.Result) -> types.ReturnTypes:
+        """Convert the response into a proper type and restore exception if any"""
+        match response.code:
+            case 200:
+                if method_properties.envelope_key not in response.result:
+                    raise exceptions.BadRequest("No data was provided in the body. Make sure to only used typed methods.")
+
+                if method_properties.return_type is None:
+                    return None
+
+                try:
+                    ta = pydantic.TypeAdapter(method_properties.return_type)
+                except common.InvalidMethodDefinition:
+                    raise exceptions.BadRequest("Typed client can only be used with typed methods.")
+
+                return ta.validate_python(response.result[method_properties.envelope_key])
+
+            case 400:
+                self._raise_exception(exceptions.BadRequest, response.result)
+
+            case 401:
+                self._raise_exception(exceptions.UnauthorizedException, response.result)
+
+            case 403:
+                self._raise_exception(exceptions.Forbidden, response.result)
+
+            case 404:
+                self._raise_exception(exceptions.NotFound, response.result)
+
+            case 409:
+                self._raise_exception(exceptions.Conflict, response.result)
+
+            case 500:
+                self._raise_exception(exceptions.ServerError, response.result)
+
+            case 503:
+                self._raise_exception(exceptions.ShutdownInProgress, response.result)
+
+            case _:
+                self._raise_exception(exceptions.ServerError, response.result)
+
+    async def _call(
+        self, method_properties: common.MethodProperties, args: list[object], kwargs: dict[str, object]
+    ) -> types.ReturnTypes:
+        """Execute a call and return the result"""
+        return self._process_response(method_properties, await self._transport_instance.call(method_properties, args, kwargs))
+
+    def __getattr__(self, name: str) -> Callable[..., Coroutine[Any, Any, types.ReturnTypes]]:
```

Comment:
why is this here? (It looks like the default behavior of python re-defined)

Reply:
I removed it